### ISSUE TITLE
MissingMethodException in LoginSilentWithCertificateAsync when upgrading from 3.13.5 to 3.19.1

### ISSUE THREAD

**Issue by tyjen** [AzureAD/azure-activedirectory-library-for-dotnet#994]
We're suddenly having issues acquiring an application token after upgrading from Microsoft.IdentityModel.Clients.ActiveDirectory 3.15.5 to 3.19.1. The last line throws the exception:

```
ClientAssertionCertificate clientAssertionCertificate = new ClientAssertionCertificate(_servicePrincipalApplicationId, _clientCertificate);
ServiceClientCredentials serviceCreds = await ApplicationTokenProvider.LoginSilentWithCertificateAsync(_subscriptionTenantId, clientAssertionCertificate);
```

`MissingMethodException: Method not found: 'System.Threading.Tasks.Task1<Microsoft.Rest.ServiceClientCredentials> Microsoft.Rest.Azure.Authentication.ApplicationTokenProvider.LoginSilentWithCertificateAsync(System.String, Microsoft.IdentityModel.Clients.ActiveDirectory.ClientAssertionCertificate)'..
`

Not sure why it's complaining it can't find that method, it clearly exists (inspected the dll myself). Is this a known issue?

**Comment by jmprieur** [comment]
thanks for reporting the issue @tyjen 
What is ``LoginSilentWithCertificateAsync`` ? I don't think it's part of ADAL.NET public API ? 
Which platform are you using? (.NET Framework? .NET core? Xamarin IOS? Xamarin Android? UWP?)

Do you have some sample code that we could use ? or repro steps?

**Comment by jennyf19** [comment]
@tyjen - we don't have an ADAL release of 3.15.5, can you double check the version number that was working previously? thanks.

**Comment by smaine** [comment]
The problematic package seems to actually be Microsoft.Rest.ClientRuntime.Azure.Authentication, which defines the ApplicationTokenProvider type. That has several API's that take in concrete instances of ClientAssertionCertificate (which no longer exists in the 3.x branch). As a result, the Net452 flavor of that package is actually incompatible with the Microsoft.IdentityModel.Clients.ActiveDirectory 3.x even though the NuSpec allows forwarding all the way to 4.0. Because they allow incompatible dependencies, it's possible to get into some really strange situations. I'll open an issue on their side as soon as I can find that codebase,

**Comment by smaine** [comment]
Some more context; ApplicationTokenProvider is defined in higher levels of the Azure SDK and provides common helper methods for authenticating to Azure. It seems to be the only place in the framework that provides any route to authenticating as a service principal using certificate credentials. Unfortunately, due to the removal of the concrete ClientAssertionCertificate in 3.x this path is broken. 

A minimal app that used client certificate credentials to authenticate as a service principal and enumerate a Key Vault will hit this issue squarely, due to the Key Vault SDK's dependency on 3.x of the ActiveDirectory client.

**Comment by jennyf19** [comment]
@smaine Thanks for the extra context. The Azure SDK is using ADAL v2. We changed the api surface in ADAL v3, which is a major version upgrade. The Azure SDK needs to update to v3.

**Comment by smaine** [comment]
In V3, where do the concrete implementations of IClientAssertionCertificate come from? Those aren't easy for non-experts to write themselves.

**Comment by jennyf19** [comment]
Try this [documentation](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/wiki/Client-credential-flows#client-credentials-with-certificate-in-adalnet) and linked code samples. 
fyi - @jmprieur

**Comment by smaine** [comment]
@jennyf19 : I'm pretty sure that sample is broken against 3.x due to this exact issue.

**Comment by jennyf19** [comment]
We'll look into it the issue further @smaine

**Comment by jmprieur** [comment]
@smaine, @tyjen let me rephrase what I understand:
- the Azure SDK leverages ADAL v2.x
- @tyjen tried to upgrade an application using th Azure SDK to ADAL v3.x, which cannot without also breaking the Azure SDK dependency as there are breaking changes between ADAL v2.x and ADAL v3.x (by design as we used [semantic versioning](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/wiki/Semantic-versioning.-API-change-management))

My question is:
- are you happy with closing this issue?
- otherwise, how can the ADAL.NET team help?

cc: @jennyf19

**Comment by jennyf19** [comment]
@smaine @tyjen 
Going to close this issue, as noted in @jmprieur's comment above. Please reach out again if you are having further issues or need help from us. Thank you.

**Comment by tyjen** [comment]
Thanks guys, we've successfully implemented a workaround and have also filed an issue on the rest client folks [here](https://github.com/Azure/autorest/issues/2853).

Our workaround was basically to avoid passing around the offending classes from the incompatible dll by implementing our own ApplicationTokenProvider, CertificateAuthenticationProvider and TokenCredentials (by looking at ADAL's source code for these classes).

**Comment by jennyf19** [comment]
@tyjen Thanks for the update.

**Comment by jmprieur** [comment]
@tyjen : are you confortable with this approach? aren't you afraid of side effects? I guess what I'm wondering is if it wouldn't be better that the SDK updates to v3.0?
